graspy/inference/nonpar.py

```diff
@@ -0,0 +1,142 @@
+# Bijan Varjavand
+# bpedigo [at] jhu.edu
+# 10.18.2018
+
+import numpy as np
+import networkx as nx
+
+from .base import BaseInference
+from ..utils import import_graph, is_symmetric, symmetrize
+from ..embed import select_dimension, AdjacencySpectralEmbed
+from sklearn.decomposition import TruncatedSVD as TSVD
+
+
+class NonparametricTest(BaseInference):
+    """
+    Two sample hypothesis test for the nonparamatric problem of determining
+    whether two random dot product graphs have the same latent positions.
+
+    Parameters
+    ----------
+    embedding : string, { 'ase' (default), 'lse', 'omnibus'}
+        String describing the embedding method to use.
+        Must be one of:
+        'ase'
+            Embed each graph separately using adjacency spectral embedding
+            and use Procrustes to align the embeddings.
+        'lse'
+            Embed each graph separately using laplacian spectral embedding
+            and use Procrustes to align the embeddings.
+        'omnibus'
+            Embed all graphs simultaneously using omnibus embedding.
+
+    n_components : None (default), or Int
+        Number of embedding dimensions. If None, the optimal embedding
+        dimensions are found by the Zhu and Godsi algorithm.
+
+    n_bootstraps : 200 (default), or Int
+        Number of bootstrap iterations.
+
+    bandwidth : None (default), of Float
+        Bandwidth to use for gaussian kernel. If None,
+        the median heuristic will be used.
+    """
+
+    def __init__(
+        self, embedding="ase", n_components=None, n_bootstraps=200, bandwidth=None
+    ):
+
+        if type(n_bootstraps) is not int:
+            raise TypeError()
+
+        if bandwidth is not None and type(bandwidth) is not float:
+            raise TypeError()
+
+        if n_bootstraps < 1:
+            msg = "{} is invalid number of bootstraps, must be greater than 1"
+            raise ValueError(msg.format(n_bootstraps))
+
+        super().__init__(embedding=embedding, n_components=n_components)
+        self.n_bootstraps = n_bootstraps
+        self.bandwidth = bandwidth
+        self.null_distribution_ = None
+        self.sample_T_statistic_ = None
+
+    def _gaussian_covariance(self, X, Y):
+        diffs = np.expand_dims(X, 1) - np.expand_dims(Y, 0)
+        if self.bandwidth == None:
+            self.bandwidth = np.median(diffs)  # TODO
+            self.bandwidth = 0.5
+        return np.exp(-0.5 * np.sum(diffs ** 2, axis=2) / self.bandwidth ** 2)
+
+    def _statistic(self, X, Y):
+        N, _ = X.shape
+        M, _ = Y.shape
+        x_stat = np.sum(self._gaussian_covariance(X, X) - np.eye(N)) / (N * (N - 1))
+        y_stat = np.sum(self._gaussian_covariance(Y, Y) - np.eye(M)) / (M * (M - 1))
+        xy_stat = np.sum(self._gaussian_covariance(X, Y)) / (N * M)
+        return x_stat - 2 * xy_stat + x_stat
```

Comment:
Should this be `x_stat - 2 * xy_stat + y_stat` not `x_stat - 2 * xy_stat + x_stat`?

Reply:
yes